RSFs (#72)
mlsa-book/MLSA: book/forests.qmd

However, when utilised in random forests, many of these parameters and steps can be safely ignored, hence this section only focuses on the components that primarily influence the resulting random forest.
This section will start by discussing decision trees and will then introduce the *bagging* algorithm used to create random forests.

### Decision Trees

*Decision trees* are a (relatively) simple machine learning model that are relatively simple to implement in software and are highly interpretable.
A decision tree takes a set of inputs and a given *splitting rule* in order to create a series of splits, or branches, in the tree that culminates in a final *leaf*, or *terminal node*.
The decision tree algorithm takes an input, a dataset, selects a variable that is used to partition the data according to some *splitting rule* into distinct non-overlapping datasets or *nodes* or *leaves*, and repeats this step for the resulting partitions, or *branches*, until some criterion has been reached.
The final nodes are referred to as *terminal nodes*.

> This is made clearer by example, (@fig-surv-ranfor) demonstrates a decision tree predicting the miles per gallon (`mpg`) of a car from the `mtcars` dataset [@datamtcars].

Reviewer: would be great to use some less overused example

During training, the algorithm identified that the first optimal variable to split the data was the `wt` variable, and that the optimal partition occurs at `wt=2.32`, i.e., one dataset is created for observations with $\code{wt} > 32$ and the other for observations with $\code{wt} \leq 2.32$.
One of the resulting nodes is left with only $7$ observations and is not split again, the other node is split according to the `disp` variable.

The graphic highlights several core features of decision trees:

> 1. They can model non-linear effects: The hierarchical structure allows for complex interactions between variables;

Reviewer: non-linearity and interactions are two different things

2. They are highly interpretable: it is easy to visualise the tree and see how predictions are made;
3. They perform variable selection: a very small subset of variables (2 out of 10) were used to train the model by carefully selecting the variables that would lead to the most heterogeneity in the resulting splits -- discussed next.

To understand how random forests work, it is worth looking a bit more into the most important components of decision trees: splitting rules, stopping rules, and terminal node predictions.

![Demonstrating classification trees using the `mtcars`  [@datamtcars] dataset and the $\pkg{party}$  [@pkgparty] package. Ovals are leaves, which indicate the variable that is being split. Edges are branches, which indicate the cut-off at which the variable is split. Rectangles are terminal nodes and include information about the number of training observations in the node and the terminal node prediction.](Figures/forests/iris_tree.png){#fig-surv-ranfor fig-alt="TODO"}

#### Splitting and Stopping Rules {.unnumbered .unlisted}

> Precisely how the data partitions/splits are derived and which variables are utilised is determined by the *splitting rule*.

Reviewer: which variables are utilised is only partially determined by the splitting rule, but also by mtry. 
maybe better to start with assuming only one variabe, explain splitting rule, than go on to generalise to multiple variables

Reviewer: hm, I guess trees use all variables as potential split variables?

Author: Eh I think it's correct to try all variables when there are few variables, mtry is just to subset to make the algorithm run faster but theoretically you use all

Reviewer: yes in trees probably. In forest it serves purposes: a) randomize the tree which improves ensemble properties, b) give variables a chance that have effect, but smaller than some dominating features with large effects (especially if some splitting rule is used that is biased towards features with large variance like Gini importance)

Author: Sure but we talk about randomisation later, so I think adding it to trees *could* be confusing?

Reviewer: agreed 👍🏻

The goal in each partition is to find two resulting leaves/nodes that have the greatest difference between them and thus the maximal homogeneity within each leaf, hence with each split, the data in each node should become increasingly similar.
The splitting rule provides a way to measure the homogeneity within the resulting nodes.
In regression, the most common splitting rule is to select a variable and cut-off (i.e., a threshold on the variable at which to separate observations) that minimises the mean squared error in the two potential resulting leaves.

For all decision tree and random forest algorithms going forward, let $L$ denote some leaf, then let $L_{xy}, L_x, L_y$ respectively be the set of observations, features, and outcomes in leaf $L$.

> Let $L_{y;i}$ be the $i$th outcome in $L_y$ and finally let $L_{\bar{y}} = \mean{L_{y;i}}$.

Reviewer: the sum is from i to n here, but should be i in L_y, i.e. cardinality of the leaf

> Let $L_{y;i}$ be the $i$th outcome in $L_y$ and finally let $L_{\bar{y}} = \mean{L_{y;i}}$.

Reviewer: sum here is from 1 to n, but should be i \in {i:i \in L_y} or something like that, i.e. cardinality of the leaf

> (j^*, c^*) = \argmin_{j, c} \sum_{y \in L^a_{y}(j,c)} (y - L^a_{\bar{Y}}(j,c))^2 + \sum_{y \in L^b_{y}(j,c)} (y - L^b_{\bar{Y}}(j,c))^2

Reviewer: ```suggestion
(j^*, c^*) = \argmin_{j, c} \sum_{y \in L^a_{y}(j,c)} (y - L^a_{\bar{y}}(j,c))^2 + \sum_{y \in L^b_{y}(j,c)} (y - L^b_{\bar{y}}(j,c))^2
```

$$ {#eq-dt-min}

This method is repeated from the first leaf to the last such that observations are included in a given leaf $L$ if they satisfy all conditions from all previous branches (splits); features may be considered multiple times in the growing process allowing complex interaction effects to be captured.

Leaves are repeatedly split until a *stopping rule* has been triggered -- i.e., a criterion that tells the algorithm to stop partitioning data.

> The** 'stopping rule' is usually a condition on the number of observations in each leaf such that leaves will continue to be split until some minimum number of observations has been reached in a leaf.

Reviewer: ```suggestion
The stopping rule is usually a condition on the number of observations in each leaf such that leaves will continue to be split until some minimum number of observations has been reached in a leaf.
```

Other conditions may be on the 'depth' of the tree, which corresponds to the number of levels of splitting, for example the tree in (@fig-surv-ranfor) has a depth of $2$.
Stopping rules are often used together, for example by setting a maximum tree depth *and* determining a minimum number of observations per leaf.
Deciding the number of minimum observations and/or the maximum depth can be performed with automated hyper-parameter optimisation.

#### Terminal Node Predictions {.unnumbered .unlisted}

The final major component of decision trees are *terminal node predictions*.
As the name suggests, this is part of the algorithm that determines how to actually make a prediction for a new observation.
A prediction is made for some new data by 'dropping' the new data 'down' the tree according to the optimal splits that were found during training.

> Returning to @fig-surv-ranfor, say a new data point is $x = \{`wt` = 3, `disp` = 250\}$, then in the first split the right branch is taken as `wt` $= 3 > 2.32$ and in the second split the left branch is taken as `disp` $= 250 \leq 258$, therefore the new data point 'lands' in the final leaf (the middle gray box).

Reviewer: `wt` and `disp` is rendered weirdly

The resulting prediction is then a simple baseline statistic computed from the training data that fell into the corresponding node.
In regression, this is most commonly (and simply) the sample mean of the training outcome data, but other terminal node predictions could be the sample median, a weighted average, or even a simple linear regression prediction from a model trained on the training data in the corresponding final leaf.
In the example above, the final prediction would be $y = 20.755$, note how the final predictions are statistics based on training data, which means all potential predictions can be saved in the original trained mode and no complex computations are required during prediction.

### Random Forests

> As is commonly the case in machine learning, higher interpretable yields poor predictive performance.

Reviewer: ```suggestion
```

Decision trees often overfit the training data, hence they have high variance, perform poorly on new data, and are not robust to even small changes in the original training data.

> Moreover, many important variables can be ignored as only a very small subset are selected, which can be particularly problematic in survival analysis when variables are often know *a priori* to have important clinical value.

Reviewer: ```suggestion
Moreover, important variables can be ignored as only subsets of dominant variables are selected for splitting.
```

Reviewer: I think it's not clear here, that this happens because all variables are available for splitting and one the ones with strongest effects are selected for splitting.

To counter these problems, *random forests* are preferred to improve prediction accuracy and decrease variance.
Random forests utilise bootstrap aggregation, or *bagging*  [@Breiman1996a], to aggregate many decision trees.
Bagging is a relatively simple algorithm, as follows:

1. **For** $b = 1,...,B$:
2. $D_b \gets \text{ Randomly sample with replacement } \dtrain$
3. $\hatg_b \gets$ \text{ Train a decision tree on } $D_b$
4. **end For**
5. $g_B \gets \{\hatg_b\}^B_{b=1}$
6. **return** $\hatg = g_B$

> Step $2$ is known as *bootstrapping*, which is the process of sampling a dataset, for example given data $\{1, 2, 3, 4\}$, standard sampling without replacement may result in the dataset $\{1, 3, 4\}$ as the first observation is drawn and removed from the original data, then the next two follow similarly.

Reviewer: this is going too much into detail, I'd just say "step 2 is known as bootstraping, i.e. sampling with replacement."

Hence, a common RSF stopping rule is the minimum number of *uncensored* observations per leaf, i.e., a leaf is not split if that would result in there being too few uncensored observations in the resulting leaves.

#### Probabilistic Predictions {.unnumbered .unlisted}

Starting with the most common survival prediction type, the algorithm requires a simple estimate for the underlying survival distribution in each of the terminal nodes, which can be estimated using the Kaplan-Meier or Nelson-Aalen methods [@Hothorn2004; @Ishwaran2008; @LeBlanc1993; @Segal1988].
By convention, the Nelson-Aalen estimator is used more commonly for RSFs and is thus focused on in this section, but the same logic follows for the Kaplan-Meier estimator.

Once the Nelson-Aalen estimate for the hazard function is estimated for individual trees, the question is then how how to combine them to create a single random forests prediction.
As Nelson-Aalen estimates produce discrete hazard functions, the same bagging algorithm can be used relatively simply by averaging the estimated hazard probability at each observed time-point [@Hothorn2004].

> Let $b_m \in \{1,...,b_M\}$ be a terminal node in a survival decision tree, $b$, with $b_M \in \PReals$ terminal nodes, then if an observation, $x$, lands in terminal node $b_m$, its predicted cumulative hazard is given by,

Reviewer: true, my bad. didn't read properly

$$
\hat{H}_b(\tau) = \sum_{\{i: i \in L^{b_m} \cap t_i \leq \tau\}} \frac{o_i}{n_i}
$$ {#eq-surv-nelson}

where $o_i$ and $n_i$ are the observed number of events, and the observations at risk, respectively at time $t_i$.
The bootstrapped Nelson-Aalen estimator is then defined as [@Ishwaran2008]

$$
\hat{H}_{Boot}(\tau) = \frac{1}{B} \sum^B_{b=1} \hat{H}_b(\tau)
$$ {#eq-surv-nelson-boot}

where $B$ is the total number of bootstrapped estimators.
The bootstrapped Kaplan-Meier estimator is calculated analogously.
More generally these can be considered as a uniform mixture of $B$ distributions (@sec-car).

> Extensions to competing risks follow naturally using bootstrapped cause-specific cumulative incidence functions.

Author: Agreed but I need help here, can I add a note for you to add detail?

> Extensions to competing risks follow naturally using bootstrapped cause-specific cumulative incidence functions.

Reviewer: ok

## Further reading

* A comprehensive review of random survival forests (RSFs) is provided in Bou-Hamad (2011) [@Bou-Hamad2011], which includes extensions to time-varying covariates and different censoring types.
* The discussion of decision trees omitted many methods for growing and pruning trees, if you are interest in those technical details see [@Breiman1984].

> * RSFs have been proven to perform well in benchmark experiments, especially on high-dimensional data, see [@Herrmann2020; @] for examples.

Author: I don't understand the comment; RSF performs better on high-dim; CPH on low-dim
